File: internal/orchestrator/skills.go
package orchestrator

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strings"

	"github.com/Z-M-Huang/openhive/internal/domain"
	"gopkg.in/yaml.v3"
)

// SkillLoader loads skill definitions from local files.
// Skill files are stored in <teamsDir>/teams/<slug>/skills/<name>.yaml (or .json).
// No HTTP fetching — local files only.
type SkillLoader struct {
	teamsDir string
	logger   *slog.Logger
}

// NewSkillLoader creates a new SkillLoader.
// teamsDir is the base directory containing teams/<slug>/ directories.
func NewSkillLoader(teamsDir string, logger *slog.Logger) *SkillLoader {
	return &SkillLoader{
		teamsDir: teamsDir,
		logger:   logger,
	}
}

// LoadSkill loads a single skill by name from a team's skills directory.
// skillName must not contain path separators or traversal components.
// Returns NotFoundError if the file does not exist.
func (s *SkillLoader) LoadSkill(teamSlug, skillName string) (*domain.Skill, error) {
	if err := domain.ValidateSlug(teamSlug); err != nil {
		return nil, err
	}
	if err := validateSkillName(skillName); err != nil {
		return nil, err
	}

	skillsDir := filepath.Join(s.teamsDir, "teams", teamSlug, "skills")

	// Try .yaml, .yml, .json, .skill.md in order
	extensions := []string{".yaml", ".yml", ".json", ".skill.md"}
	var data []byte
	var ext string
	var readErr error
	for _, e := range extensions {
		path := filepath.Join(skillsDir, skillName+e)
		data, readErr = os.ReadFile(path)
		if readErr == nil {
			ext = e
			break
		}
	}

	if readErr != nil {
		return nil, &domain.NotFoundError{
			Resource: "skill",
			ID:       teamSlug + "/" + skillName,
		}
	}

	skill, parseErr := parseSkillFile(data, ext)
	if parseErr != nil {
		return nil, fmt.Errorf("failed to parse skill %s: %w", skillName, parseErr)
	}

	// Override name with the file name (canonical)
	if skill.Name == "" {
		skill.Name = skillName
	}

	if err := ValidateSkill(skill); err != nil {
		return nil, fmt.Errorf("skill %s validation failed: %w", skillName, err)
	}

	return skill, nil
}

// LoadAllSkills loads all skills from a team's skills directory.
func (s *SkillLoader) LoadAllSkills(teamSlug string) ([]domain.Skill, error) {
	if err := domain.ValidateSlug(teamSlug); err != nil {
		return nil, err
	}

	skillsDir := filepath.Join(s.teamsDir, "teams", teamSlug, "skills")

	entries, err := os.ReadDir(skillsDir)
	if err != nil {
		if os.IsNotExist(err) {
			return []domain.Skill{}, nil
		}
		return nil, fmt.Errorf("failed to read skills directory: %w", err)
	}

	var skills []domain.Skill
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		name := entry.Name()

		// Detect extension: check for compound .skill.md first, then single extensions.
		var ext, skillName string
		if strings.HasSuffix(name, ".skill.md") {
			ext = ".skill.md"
			skillName = strings.TrimSuffix(name, ".skill.md")
		} else {
			ext = filepath.Ext(name)
			if ext != ".yaml" && ext != ".yml" && ext != ".json" {
				continue
			}
			skillName = strings.TrimSuffix(name, ext)
		}

		// Exact extension check: the skill base name (filename minus the single
		// recognised extension) must be a valid skill identifier. This rejects
		// compound filenames like 'skill.yaml.json' where the derived name
		// 'skill.yaml' contains an illegal character.
		if err := validateSkillName(skillName); err != nil {
			s.logger.Warn("skipping skill file with invalid base name",
				"team", teamSlug,
				"file", name,
				"reason", err.Error(),
			)
			continue
		}

		skill, loadErr := s.LoadSkill(teamSlug, skillName)
		if loadErr != nil {
			s.logger.Warn("failed to load skill", "team", teamSlug, "skill", skillName, "error", loadErr)
			continue
		}
		skills = append(skills, *skill)
	}

	return skills, nil
}

// ValidateSkill validates a skill definition.
// Checks required fields, model_tier, and tool names.
func ValidateSkill(skill *domain.Skill) error {
	if skill.Name == "" {
		return &domain.ValidationError{Field: "name", Message: "skill name is required"}
	}
	if err := validateSkillName(skill.Name); err != nil {
		return err
	}
	if skill.ModelTier != "" {
		if _, err := domain.ParseModelTier(skill.ModelTier); err != nil {
			return &domain.ValidationError{
				Field:   "model_tier",
				Message: fmt.Sprintf("invalid model_tier: %s (must be haiku, sonnet, or opus)", skill.ModelTier),
			}
		}
	}
	// Validate tool names against known tools
	for _, toolName := range skill.Tools {
		if err := validateToolName(toolName); err != nil {
			return &domain.ValidationError{
				Field:   "tools",
				Message: fmt.Sprintf("invalid tool name %q: %s", toolName, err.Error()),
			}
		}
	}
	return nil
}

// validateSkillName validates a skill name (must be a safe identifier, not a path).
func validateSkillName(name string) error {
	if name == "" {
		return &domain.ValidationError{Field: "skill_name", Message: "skill name cannot be empty"}
	}
	if strings.Contains(name, "..") {
		return &domain.ValidationError{Field: "skill_name", Message: "skill name must not contain '..' (path traversal)"}
	}
	if strings.ContainsAny(name, "/\\") {
		return &domain.ValidationError{Field: "skill_name", Message: "skill name must not contain path separators"}
	}
	// Allow letters, digits, hyphens, underscores
	for _, c := range name {
		if !((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '-' || c == '_') {
			return &domain.ValidationError{
				Field:   "skill_name",
				Message: fmt.Sprintf("skill name contains invalid character: %q", c),
			}
		}
	}
	return nil
}

// knownToolNames is the set of tool names registered by the orchestrator.
// validateToolName is permissive — tool names starting with known prefixes are allowed.
func validateToolName(name string) error {
	if name == "" {
		return fmt.Errorf("tool name cannot be empty")
	}
	// Tools may only contain alphanumerics and underscores
	for _, c := range name {
		if !((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '_') {
			return fmt.Errorf("tool name %q contains invalid character %q", name, c)
		}
	}
	return nil
}

// parseSkillFile parses skill YAML or JSON data.
func parseSkillFile(data []byte, ext string) (*domain.Skill, error) {
	var skill domain.Skill
	switch strings.ToLower(ext) {
	case ".yaml", ".yml":
		if err := yaml.Unmarshal(data, &skill); err != nil {
			return nil, fmt.Errorf("YAML parse error: %w", err)
		}
	case ".json":
		if err := json.Unmarshal(data, &skill); err != nil {
			return nil, fmt.Errorf("JSON parse error: %w", err)
		}
	case ".skill.md":
		parsed, err := parseSkillMarkdown(data)
		if err != nil {
			return nil, fmt.Errorf("markdown parse error: %w", err)
		}
		skill = *parsed
	default:
		return nil, fmt.Errorf("unsupported skill file extension: %s", ext)
	}
	return &skill, nil
}

// parseSkillMarkdown parses a .skill.md file with optional YAML frontmatter.
// Format:
//
//	---
//	name: my-skill
//	description: A useful skill
//	model_tier: sonnet
//	tools: [tool1, tool2]
//	---
//	Body text becomes system_prompt_addition.
func parseSkillMarkdown(data []byte) (*domain.Skill, error) {
	content := string(data)
	var skill domain.Skill

	// Check for YAML frontmatter delimited by "---"
	if strings.HasPrefix(strings.TrimSpace(content), "---") {
		trimmed := strings.TrimSpace(content)
		// Find closing "---"
		rest := trimmed[3:] // skip opening "---"
		idx := strings.Index(rest, "\n---")
		if idx >= 0 {
			frontmatter := rest[:idx]
			body := strings.TrimSpace(rest[idx+4:]) // skip "\n---"
			if err := yaml.Unmarshal([]byte(frontmatter), &skill); err != nil {
				return nil, fmt.Errorf("frontmatter parse error: %w", err)
			}
			skill.SystemPromptAddition = body
		} else {
			// No closing delimiter: treat entire content as system prompt
			skill.SystemPromptAddition = strings.TrimSpace(content)
		}
	} else {
		// No frontmatter: entire file is the system prompt addition
		skill.SystemPromptAddition = strings.TrimSpace(content)
	}

	return &skill, nil
}

File: backend/src/orchestrator/skills.ts
/**
 * OpenHive Backend - SkillLoader
 *
 * Loads skill definitions from the workspace's .claude/skills/ directory.
 * Each skill lives in its own subdirectory: .claude/skills/<name>/SKILL.md
 *
 * No HTTP fetching — local files only.
 */

import { type Dirent, readFileSync, readdirSync } from 'node:fs';
import { join } from 'node:path';
import { parse as parseYaml } from 'yaml';

import type { Skill } from '../domain/types.js';
import { NotFoundError, ValidationError } from '../domain/errors.js';
import { parseModelTier } from '../domain/enums.js';

// ---------------------------------------------------------------------------
// Logger interface — minimal structured logger required by SkillLoader
// ---------------------------------------------------------------------------

/**
 * Minimal structured logger interface required by SkillLoader.
 * Compatible with pino or any standard structured logger.
 */
export interface SkillLoaderLogger {
  warn(msg: string, data?: Record<string, unknown>): void;
}

// ---------------------------------------------------------------------------
// SkillLoader
// ---------------------------------------------------------------------------

/**
 * Loads skill definitions from a team workspace's .claude/skills/ directory.
 *
 * Skill files are stored under:
 *   <workspaceBase>/.claude/skills/<name>/SKILL.md
 *
 * The workspaceBase is the root of a team's workspace directory
 * (e.g. .run/workspace/teams/<slug>/ or .run/workspace/ for main).
 */
export class SkillLoader {
  private readonly workspaceBase: string;
  private readonly logger: SkillLoaderLogger;

  constructor(workspaceBase: string, logger: SkillLoaderLogger) {
    this.workspaceBase = workspaceBase;
    this.logger = logger;
  }

  // -------------------------------------------------------------------------
  // loadSkill
  // -------------------------------------------------------------------------

  /**
   * Loads a single skill by name from the workspace's .claude/skills/ directory.
   *
   * Reads: <workspaceBase>/.claude/skills/<skillName>/SKILL.md
   *
   * skillName must not contain path separators or traversal components.
   *
   * Throws:
   *   - ValidationError if skillName is invalid
   *   - NotFoundError if the skill directory or SKILL.md does not exist
   *   - Error if the skill file cannot be parsed or fails validation
   */
  loadSkill(skillName: string): Skill {
    validateSkillName(skillName);

    const skillFile = join(this.workspaceBase, '.claude', 'skills', skillName, 'SKILL.md');

    let data: Buffer;
    try {
      data = readFileSync(skillFile);
    } catch (err) {
      if (isEnoent(err)) {
        throw new NotFoundError('skill', skillName);
      }
      throw err;
    }

    let skill: Skill;
    try {
      skill = parseSkillMarkdown(data.toString('utf8'));
    } catch (err) {
      throw new Error(
        `failed to parse skill ${skillName}: ${err instanceof Error ? err.message : String(err)}`,
      );
    }

    if (!skill.name || skill.name === '') {
      skill = { ...skill, name: skillName };
    }

    validateSkill(skill);

    return skill;
  }

  // -------------------------------------------------------------------------
  // loadAllSkills
  // -------------------------------------------------------------------------

  /**
   * Loads all skills from the workspace's .claude/skills/ directory.
   *
   * Each skill is expected in its own subdirectory: .claude/skills/<name>/SKILL.md
   *
   * Returns an empty array if the skills directory does not exist.
   * Logs and skips any skill that fails to load.
   *
   * Throws:
   *   - Error if the skills directory cannot be read (other than ENOENT)
   */
  loadAllSkills(): Skill[] {
    const skillsDir = join(this.workspaceBase, '.claude', 'skills');

    let entries: Dirent<string>[];
    try {
      entries = readdirSync(skillsDir, { withFileTypes: true }) as Dirent<string>[];
    } catch (err) {
      if (isEnoent(err)) {
        return [];
      }
      throw new Error(
        `failed to read skills directory: ${err instanceof Error ? err.message : String(err)}`,
      );
    }

    const skills: Skill[] = [];

    for (const entry of entries) {
      // Only process directories — each skill lives in its own subdirectory.
      if (!entry.isDirectory()) {
        continue;
      }

      const skillName = entry.name;

      try {
        validateSkillName(skillName);
      } catch (err) {
        this.logger.warn('skipping skill directory with invalid base name', {
          dir: skillName,
          reason: err instanceof Error ? err.message : String(err),
        });
        continue;
      }

      let skill: Skill;
      try {
        skill = this.loadSkill(skillName);
      } catch (err) {
        this.logger.warn('failed to load skill', {
          skill: skillName,
          error: err instanceof Error ? err.message : String(err),
        });
        continue;
      }

      skills.push(skill);
    }

    return skills;
  }
}

// ---------------------------------------------------------------------------
// validateSkill
// ---------------------------------------------------------------------------

/**
 * Validates a skill definition.
 *
 * Throws ValidationError if any field is invalid.
 */
export function validateSkill(skill: Skill): void {
  if (!skill.name || skill.name === '') {
    throw new ValidationError('name', 'skill name is required');
  }

  validateSkillName(skill.name);

  if (skill.model_tier !== undefined && skill.model_tier !== '') {
    try {
      parseModelTier(skill.model_tier);
    } catch {
      throw new ValidationError(
        'model_tier',
        `invalid model_tier: ${skill.model_tier} (must be haiku, sonnet, or opus)`,
      );
    }
  }

  if (skill.tools !== undefined) {
    for (const toolName of skill.tools) {
      try {
        validateToolName(toolName);
      } catch (err) {
        throw new ValidationError(
          'tools',
          `invalid tool name ${JSON.stringify(toolName)}: ${err instanceof Error ? err.message : String(err)}`,
        );
      }
    }
  }
}

// ---------------------------------------------------------------------------
// validateSkillName (internal)
// ---------------------------------------------------------------------------

/**
 * Validates that a skill name is safe and well-formed.
 *
 * Rejects empty strings, path traversal components, path separators,
 * and characters outside [a-zA-Z0-9_-].
 *
 * Throws ValidationError if invalid.
 */
export function validateSkillName(name: string): void {
  if (name === '') {
    throw new ValidationError('skill_name', 'skill name cannot be empty');
  }
  if (name.includes('..')) {
    throw new ValidationError('skill_name', "skill name must not contain '..' (path traversal)");
  }
  if (name.includes('/') || name.includes('\\')) {
    throw new ValidationError('skill_name', 'skill name must not contain path separators');
  }
  for (const c of name) {
    const cp = c.codePointAt(0)!;
    const isLower = cp >= 97 && cp <= 122; // a-z
    const isUpper = cp >= 65 && cp <= 90; // A-Z
    const isDigit = cp >= 48 && cp <= 57; // 0-9
    const isHyphen = c === '-';
    const isUnderscore = c === '_';
    if (!isLower && !isUpper && !isDigit && !isHyphen && !isUnderscore) {
      throw new ValidationError(
        'skill_name',
        `skill name contains invalid character: ${JSON.stringify(c)}`,
      );
    }
  }
}

// ---------------------------------------------------------------------------
// validateToolName (internal)
// ---------------------------------------------------------------------------

/**
 * Validates that a tool name contains only alphanumerics and underscores.
 *
 * Throws an Error (not ValidationError) if invalid — the caller
 * (validateSkill) wraps it into a ValidationError.
 */
function validateToolName(name: string): void {
  if (name === '') {
    throw new Error('tool name cannot be empty');
  }
  for (const c of name) {
    const cp = c.codePointAt(0)!;
    const isLower = cp >= 97 && cp <= 122; // a-z
    const isUpper = cp >= 65 && cp <= 90; // A-Z
    const isDigit = cp >= 48 && cp <= 57; // 0-9
    const isUnderscore = c === '_';
    if (!isLower && !isUpper && !isDigit && !isUnderscore) {
      throw new Error(`tool name ${JSON.stringify(name)} contains invalid character ${JSON.stringify(c)}`);
    }
  }
}

// ---------------------------------------------------------------------------
// parseSkillMarkdown (internal)
// ---------------------------------------------------------------------------

/**
 * Parses a SKILL.md file.
 *
 * If the file starts with a YAML front-matter block (--- ... ---), parses
 * the front-matter as YAML into a Skill and uses the remaining body as
 * system_prompt_addition. If no front-matter is present, the entire content
 * becomes system_prompt_addition.
 *
 * Uses CORE_SCHEMA for safe YAML parsing (no implicit type coercions).
 */
function parseSkillMarkdown(content: string): Skill {
  const trimmed = content.trimStart();

  if (trimmed.startsWith('---')) {
    const rest = trimmed.slice(3); // skip opening '---'
    const idx = rest.indexOf('\n---');
    if (idx >= 0) {
      const frontmatter = rest.slice(0, idx);
      const body = rest.slice(idx + 4).trimStart(); // skip '\n---'

      let parsed: unknown;
      try {
        parsed = parseYaml(frontmatter, { schema: 'core' });
      } catch (err) {
        throw new Error(
          `frontmatter parse error: ${err instanceof Error ? err.message : String(err)}`,
        );
      }

      const skill = coerceToSkill(parsed);
      skill.system_prompt_addition = body;
      return skill;
    }
  }

  // No front-matter — entire content is the system prompt addition.
  return {
    name: '',
    system_prompt_addition: trimmed,
  };
}

// ---------------------------------------------------------------------------
// coerceToSkill (internal)
// ---------------------------------------------------------------------------

/**
 * Coerces an unknown parsed value to a Skill.
 *
 * Validates that the parsed value is a plain object and extracts all known
 * Skill fields with their expected types. Unknown fields are silently ignored.
 *
 * Throws an Error if the parsed value is not a plain object.
 */
function coerceToSkill(parsed: unknown): Skill {
  if (parsed === null || typeof parsed !== 'object' || Array.isArray(parsed)) {
    throw new Error('skill file must contain a mapping object at the top level');
  }

  const obj = parsed as Record<string, unknown>;

  const skill: Skill = {
    name: typeof obj['name'] === 'string' ? obj['name'] : '',
  };

  if (typeof obj['description'] === 'string') {
    skill.description = obj['description'];
  }
  if (typeof obj['model_tier'] === 'string') {
    skill.model_tier = obj['model_tier'];
  }
  if (Array.isArray(obj['tools'])) {
    skill.tools = obj['tools'].filter((t): t is string => typeof t === 'string');
  }
  if (typeof obj['system_prompt_addition'] === 'string') {
    skill.system_prompt_addition = obj['system_prompt_addition'];
  }

  return skill;
}

// ---------------------------------------------------------------------------
// isEnoent (internal)
// ---------------------------------------------------------------------------

/**
 * Returns true if the given error represents a file-not-found (ENOENT) condition.
 */
function isEnoent(err: unknown): boolean {
  return (
    err !== null &&
    typeof err === 'object' &&
    'code' in err &&
    (err as { code: unknown }).code === 'ENOENT'
  );
}
